### PR TITLE
Add Materials in Course Page, add link between Course Page and Quiz Feature & Add new firebase functions

### PR DESCRIPTION
**Pull request by FlorianDinant**
## Description

This Pull Request introduces several enhancements and new features to the Course Page, improving its functionality, maintainability, and scalability. The goal is to enable dynamic management of course-related data, reduce hard-coded dependencies, and leverage Firebase for backend operations. These changes also include rigorous testing to ensure high reliability.

### Dynamic Course Description
- Implemented a dedicated section to dynamically display the course description at the top of the Course Page.
- **Source of Data:**
   - Course descriptions are now fetched from Firebase in real-time, allowing admins to update course details without modifying the front-end code.
   - Ensured proper data validation and fallback handling in case of missing or invalid data.

### Materials for Defined Time Periods
- Introduced functionality to associate materials (e.g., slides, exercises, feedback, and additional resources) with specific time periods in the course schedule.
- **Dynamic Handling:**
   - Materials are dynamically fetched and displayed based on the current or selected time period.
   - Period-specific filtering (for future materials) and sorting, ensures only relevant materials are accessible to users.
- **Backend Integration:**
   - Time period and material metadata are managed via Firebase.
   - The model of the `Material`s is available in `model/src/school/course`
- **UI Changes:**
   - Created an intuitive timeline-based interface to allow users to see by clicking on a button the future materials i.e. the materials for which the starting period is not arrived yet. Also, the materials are stored from the more recent one to the oldest one, from top to bottom.
   - Added visual cues (e.g., highlights or icons) to differentiate material types (e.g., slides vs. exercises).
   - Firebase functions have been added to ensure materials can be added/updated directly in Firebase without hardcoding. Note that a future PR will give the possibility to the teachers assigned to each course to directly modify the courses informations from the app.

### Firebase Cloud Functions
- Developed and deployed custom Firebase Cloud Functions to handle backend operations efficiently.
- **Implemented Functions:**
   - Create/Update/Delete Operations: Simplified CRUD operations for course descriptions and materials and any other data related to course (Assistants, name, assignments etc.).
- **Validation and Authorization:** 
   - Ensured all updates are securely validated with role-based access control (e.g., admin-only updates / professor-only).
- **Error Handling:**
   - Implemented error logging to Firebase Crashlytics for monitoring and debugging backend issues.

### Comprehensive Testing
- Created extensive test coverage for the new features to ensure system stability and reliability.
- **Test Coverage:**
- **Unit Tests:** Validated all front-end components and backend Firebase functions.
- **Integration Tests:** Tested end-to-end data flow from Firebase to the Course Page.
- **UI Tests:** Ensured the Course Page renders correctly under various scenarios (e.g., with and without materials or descriptions).
- **Metrics:**
   - Achieved >80% test coverage for the Course Page and related components.
   - Used Jest and React Testing Library for front-end tests.

## Additional Enhancements
- Refactored related components to improve modularity and reusability.
- Linked the CoursePage to the Quiz page when a quiz assignment is clicked (waiting for @FerdinandGomerChiu PR to ensure it works correctly)

## Checklist
✅ Dynamic course description rendering with real-time Firebase integration.
✅ Material display organized by defined time periods.
✅ Firebase Cloud Functions deployed for secure and efficient data operations.
✅ Unit, integration, and UI tests with >80% coverage.

### Course Description:
- Add a new course description in Firebase and verify it is displayed correctly on the Course Page.
- Test edge cases, such as missing or malformed descriptions, to confirm proper fallback handling.
### Materials for Time Periods:
- Add various materials in Firebase, associating them with specific time periods.
- Verify the correct materials are displayed for each time period.
- Test switching between periods to ensure data updates dynamically.
### Testing Suite:
- Run all tests (npm test) and ensure no errors or warnings occur.
- Confirm >80% coverage in the test summary.

## Screens :
Screenshot 2024-11-22 at 01 45 28
Screenshot 2024-11-22 at 01 46 11
Screenshot 2024-11-22 at 01 46 39

### CODE REVIEW

File: edweiss-app/components/courses/MaterialDisplay.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ * This component is responsible for displaying a material in the course page.
+ * 
+ * @param item - The material data to be displayed.
+ * @param key - The key of the assignment in the list.
+ * 
+ * @returns JSX.Element - The rendered component for the assignment display.
+ */
+const MaterialDisplay: ReactComponent<{ item: Material; }> = ({ item }) => {
+
+    const formatDateRange = (fromSeconds: number, toSeconds: number) => {
```
[the remark this answers is not in the document]
Comment (FlorianDinant) [reply edweiss-app/components/courses/MaterialDisplay.tsx:47]: Hello! Yes there is a reason why formatterOptions doesn't define `year: 'numeric'` it is because I didn't wanted to repeat 2 times the same year e.g. display `Sept 9 - Sept 15 2024` instead of `Sept 9 2024 - Sept 15 2024`

[the remark this answers is not in the document]
Comment (holoisme) [reply edweiss-app/components/courses/MaterialDisplay.tsx:52]: Does specifying the year really necessary? I can hardly find a use for it. It just seems to take more space

File: edweiss-app/components/courses/AssignmentDisplay.tsx [hunk 2]
```diff
@@ -48,22 +49,22 @@ export type AssignmentWithColor = Assignment & { color: Color; };
...
+            <TText testID={testIDs.addToDoText} size={16} bold color='constantWhite'>{t(`course:add_to_todo`)}</TText>
         </TView>
     );
 
     const assignmentView = () => (
-        <TView testID={testIDs.assignmentView} flexDirection='row' alignItems="center" justifyContent='space-between' key={key}>
+        <TView testID={testIDs.assignmentView} flexDirection='row' alignItems="center" justifyContent='space-between'>
             {/* TODO: Handle onPress event qui envoie vers le quiz ou la soumission. ATTENTION SI LE QUIZ OU SUBMIT EST DEJA FINI */}
-            <TTouchableOpacity testID={testIDs.assignmentTouchable} backgroundColor='mantle' flexDirection='row' alignItems='center' py={12} bb={1} borderColor='crust' flex={1}>
+            <TTouchableOpacity testID={testIDs.assignmentTouchable} onPress={item.type === 'quiz' ? () => router.push({ pathname: `/(app)/quiz/temporaryQuizStudentView`, params: { quizId: id, courseId: courseID } }) : undefined} backgroundColor='mantle' flexDirection='row' alignItems='center' py={12} bb={1} borderColor='crust' flex={1}>
```
[the remark this answers is not in the document]
Comment (FlorianDinant) [reply edweiss-app/components/courses/AssignmentDisplay.tsx]: The modification to add is `disabled={item.type !== 'quiz'}` but yeah thank you very much you are right to suggest me that !

File: edweiss-app/components/courses/MaterialDisplay.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (FlorianDinant) [reply edweiss-app/components/courses/MaterialDisplay.tsx:52]: Yeah I agree I will do that!

**Review by holoisme — APPROVED** [review]
Alright, seems all good to me! :+1:
